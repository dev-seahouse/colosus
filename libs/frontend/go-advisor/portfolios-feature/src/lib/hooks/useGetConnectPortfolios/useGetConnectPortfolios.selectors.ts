import { useQuery } from '@tanstack/react-query';
import { useCallback } from 'react';

import { getConnectPortfoliosQuery } from './useGetConnectPortfolios';
import type { GetConnectPortfoliosData } from './useGetConnectPortfolios';

/**
 * hoook to return available portfolios
 */
export const useSelectConnectPortfolios = () => {
  return useQuery({
    ...getConnectPortfoliosQuery(),
    select: useCallback((data: GetConnectPortfoliosData) => {
      if (!data || !data.portfolioSummaries) {
        return [];
      }

      return data.portfolioSummaries ?? [];
    }, []),
  });
};
