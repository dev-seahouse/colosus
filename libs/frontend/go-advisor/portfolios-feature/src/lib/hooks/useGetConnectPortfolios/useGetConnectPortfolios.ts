import { useQuery } from '@tanstack/react-query';
import { ConnectAdvisorPortfolioSummaryApi } from '@bambu/api-client';
import type { ConnectAdvisorGetPortfoliosResponseDto } from '@bambu/api-client';
import type { QueryClient } from '@tanstack/react-query';

export type GetConnectPortfoliosData = ConnectAdvisorGetPortfoliosResponseDto;

export const getConnectPortfoliosQuery = () => ({
  queryKey: ['getPortfolios'],
  queryFn: async () => {
    const connectAdvisorPortfolioSummaryApi =
      new ConnectAdvisorPortfolioSummaryApi();
    const res = await connectAdvisorPortfolioSummaryApi.getPortfolios();

    return res.data;
  },
});

export const getConnectPortfoliosLoader =
  (queryClient: QueryClient) => async (): Promise<GetConnectPortfoliosData> => {
    const query = getConnectPortfoliosQuery();

    // ⬇️ return data or fetch it
    return (
      queryClient.getQueryData(query.queryKey) ??
      (await queryClient.fetchQuery(query))
    );
  };

export interface UseGetConnectPortfoliosOptions {
  initialData?: Partial<GetConnectPortfoliosData>;
}
export function useGetConnectPortfolios({
  initialData,
}: UseGetConnectPortfoliosOptions = {}) {
  return useQuery({
    ...getConnectPortfoliosQuery(),
    initialData,
  });
}

export default useGetConnectPortfolios;
