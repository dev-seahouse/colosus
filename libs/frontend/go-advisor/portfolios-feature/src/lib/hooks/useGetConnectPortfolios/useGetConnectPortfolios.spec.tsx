import {
  renderHook,
  queryClientWrapper,
  waitFor,
} from '@bambu/react-test-utils';

import useGetConnectPortfolios from './useGetConnectPortfolios';

describe('useGetConnectPortfolios', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetConnectPortfolios(), {
      wrapper: queryClientWrapper,
    });

    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data).toMatchSnapshot();
  });
});
