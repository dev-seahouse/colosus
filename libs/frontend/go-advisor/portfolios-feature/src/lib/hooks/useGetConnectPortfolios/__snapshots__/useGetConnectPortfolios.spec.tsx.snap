// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetConnectPortfolios > should render successfully 1`] = `
{
  "portfolioSummaries": [
    {
      "assetClassAllocation": [
        {
          "assetClass": "Equity",
          "included": true,
          "percentOfPortfolio": "3.5",
        },
      ],
      "description": "This portfolio primarily looks to preserve capital through a conservative asset allocation.",
      "expectedReturnPercent": "4",
      "expectedVolatilityPercent": "3.5",
      "key": "CONSERVATIVE",
      "name": "Conservative Portfolio",
      "reviewed": true,
      "riskProfileId": "b6739bd8-6267-4bf7-bb1f-e112c54bec71",
      "showSummaryStatistics": true,
    },
  ],
}
`;
